examples/plot_metric_examples.py

```diff
@@ -0,0 +1,410 @@
+    plt.clf()
+
+    tsne = TSNE()
+    X_embedded = tsne.fit_transform(X)
+    plt.scatter(X_embedded[:, 0], X_embedded[:, 1], c=Y, cmap=plt.cm.Paired)
+
+    plt.xticks(())
+    plt.yticks(())
+
+    plt.show()
+
+###################################
+# Let's now plot the dataset as is.
+
+plot(X, Y)
+
+
+######################################################################
+# Metric Learning
+# ^^^^^^^^^^^^^^^
+# 
+# Why is Metric Learning useful? We can, with prior knowledge of which
+# points are supposed to be closer, figure out a better way to understand
+# distances between points. Especially in higher dimensions when Euclidean
+# distances are a poor way to measure distance, this becomes very useful.
+# 
+# Basically, we learn this distance:
+# :math:`D(x,y)=\sqrt{(x-y)\,M^{-1}(x-y)}`. And we learn this distance by
+# learning a Matrix :math:`M`, based on certain constraints.
+# 
+# Some good reading material for the same can be found
+# `here <https://arxiv.org/pdf/1306.6709.pdf>`__. It serves as a good
+# literature review of Metric Learning.
+# 
+# We will briefly explain the metric-learning algorithms implemented by
+# metric-learn, before providing some examples for it's usage, and also
+# discuss how to go about doing manual constraints.
+# 
+# Metric-learn can be easily integrated with your other machine learning
+# pipelines, and follows (for the most part) scikit-learn conventions.
+# 
+
+
+######################################################################
+# Large Margin Nearest Neighbour
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+# 
+# LMNN is a metric learning algorithm primarily designed for k-nearest
+# neighbor classification. The algorithm is based on semidefinite
+# programming, a sub-class of convex programming (as most Metric Learning
+# algorithms are).
+# 
+# The main intuition behind LMNN is to learn a pseudometric under which
+# all data instances in the training set are surrounded by at least k
+# instances that share the same class label. If this is achieved, the
+# leave-one-out error (a special case of cross validation) is minimized.
+# You'll notice that the points from the same labels are closer together,
+# but they are not necessary in a same cluster. This is particular to LMNN
+# and we'll see that some other algorithms implicitly enforce points from
+# the same class to cluster together.
+# 
+# You can find the paper
+# `here <http://jmlr.csail.mit.edu/papers/volume10/weinberger09a/weinberger09a.pdf>`__.
+# 
+
+
+######################################################################
+# Fit and then transform!
+# -----------------------
+# 
+
+# setting up LMNN
+lmnn = metric_learn.LMNN(k=5, learn_rate=1e-6)
+
+# fit the data!
+lmnn.fit(X, Y)
+
+# transform our input space
+X_lmnn = lmnn.transform(X)
+
+
+######################################################################
+# So what have we learned? The matrix :math:`M` we talked about before.
+# Let's see what it looks like.
+# 
+
+lmnn.metric()
```

Comment:
does not print in the doc. should use `print`? but if it is not possible to easily interpret what we see, we can get rid of it